src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+                          select nodeInCurrent;
+            return matches;
+        }
+
+        private static bool NodeMatchesExpression(SemanticModel originalSemanticModel,
+            TExpressionSyntax expressionInOriginal, TExpressionSyntax nodeInCurrent, bool allOccurrences,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            if (nodeInCurrent == expressionInOriginal)
+            {
+                return true;
+            }
+
+            if (!allOccurrences)
+            {
+                return false;
+            }
+            else
+            {
+                // Original expression and current node being semantically equivalent isn't enough when the original expression 
+                // is a member access via instance reference (either implicit or explicit), the check only ensures that the expression
+                // and current node are both backed by the same member symbol. So in this case, in addition to SemanticEquivalence check, 
+                // we also check if expression and current node are both instance member access.
+                //
+                // For example, even though the first `c` binds to a field and we are introducing a local for it,
+                // we don't want other references to that field to be replaced as well (i.e. the second `c` in the expression).
+                //
+                //  class C
+                //  {
+                //      C c;
+                //      void Test()
+                //      {
+                //          var x = [|c|].c;
+                //      }
+                //  }
+
+                var originalOperation = originalSemanticModel.GetOperation(expressionInOriginal, cancellationToken);
+                if (originalOperation != null && IsInstanceMemberReference(originalOperation))
+                {
+                    var currentOperation = originalSemanticModel.GetOperation(nodeInCurrent, cancellationToken);
+                    return currentOperation != null && IsInstanceMemberReference(currentOperation) && SemanticEquivalence.AreEquivalent(
+                        originalSemanticModel, originalSemanticModel, expressionInOriginal, nodeInCurrent);
+                }
+
+                return SemanticEquivalence.AreEquivalent(
+                    originalSemanticModel, originalSemanticModel, expressionInOriginal, nodeInCurrent);
+            }
+
+            static bool IsInstanceMemberReference(IOperation operation)
+                => operation is IMemberReferenceOperation memberReferenceOperation &&
+                    memberReferenceOperation.Instance?.Kind == OperationKind.InstanceReference;
+        }
+
+        private static string CreateDisplayText(TExpressionSyntax expression, ISyntaxFactsService syntaxFacts, string resourceString)
```

Comment:
move this to callsite.  do this once.  then reference this value from all the palces that need it instead of recomputing it 6 times.

Reply:
fixed